pkg/codegen/hcl2/model/type_union.go

```diff
@@ -32,38 +33,69 @@ type UnionType struct {
 	Annotations []interface{}
 
 	s string
+	h uint32
+}
+
+func containsType(t Type, types []Type) bool {
+	for _, u := range types {
+		if t.Equals(u) {
+			return true
+		}
+	}
+	return false
 }
 
 // NewUnionTypeAnnotated creates a new union type with the given element types and annotations.
 // Any element types that are union types are replaced with their element types.
 func NewUnionTypeAnnotated(types []Type, annotations ...interface{}) Type {
-	var elementTypes []Type
+	lenTypes := 0
 	for _, t := range types {
-		if union, isUnion := t.(*UnionType); isUnion {
-			elementTypes = append(elementTypes, union.ElementTypes...)
+		if u, isUnion := t.(*UnionType); isUnion {
+			lenTypes += len(u.ElementTypes)
 		} else {
-			elementTypes = append(elementTypes, t)
+			lenTypes++
 		}
 	}
 
-	sort.Slice(elementTypes, func(i, j int) bool {
-		return elementTypes[i].String() < elementTypes[j].String()
-	})
-
-	dst := 0
-	for src := 0; src < len(elementTypes); {
-		for src < len(elementTypes) && elementTypes[src].Equals(elementTypes[dst]) {
-			src++
+	elementTypes := make([]Type, 0, lenTypes)
+	if lenTypes < 16 {
+		for _, t := range types {
+			if union, ok := t.(*UnionType); ok {
+				for _, t := range union.ElementTypes {
+					if !containsType(t, elementTypes) {
+						elementTypes = append(elementTypes, t)
+					}
+				}
+			} else {
+				if !containsType(t, elementTypes) {
+					elementTypes = append(elementTypes, t)
+				}
+			}
 		}
-		dst++
-
-		if src < len(elementTypes) {
-			elementTypes[dst] = elementTypes[src]
+	} else {
+		var tt typeTable
+		for _, t := range types {
+			if union, isUnion := t.(*UnionType); isUnion {
+				for _, t := range union.ElementTypes {
+					if _, ok := tt.lookup(t, nil); !ok {
+						tt.insert(t, nil)
+						elementTypes = append(elementTypes, t)
+					}
+				}
+			} else {
+				if _, ok := tt.lookup(t, nil); !ok {
+					tt.insert(t, nil)
+					elementTypes = append(elementTypes, t)
+				}
+			}
 		}
 	}
-	elementTypes = elementTypes[:dst]
 
 	if len(elementTypes) == 1 {
+		tt := reflect.ValueOf(elementTypes[0])
+		if !tt.Elem().IsValid() {
+			panic("what?")
```

Comment:
Just to be on the safe side let's put a bit more descriptive string into the panic :)

Reply:
oh, oops--this was debugging code that should be removed 😬